python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,178 @@
+    """ID of the “current” spec that writers should use by default."""
+
+    last_partition_id: int = Field(alias="last-partition-id")
+    """An integer; the highest assigned partition field ID across all
+    partition specs for the table. This is used to ensure partition fields
+    are always assigned an unused ID when evolving specs."""
+
+    properties: dict
+    """	A string to string map of table properties. This is used to
+    control settings that affect reading and writing and is not intended
+    to be used for arbitrary metadata. For example, commit.retry.num-retries
+    is used to control the number of commit retries."""
+
+    current_snapshot_id: int = Field(alias="current-snapshot-id")
+    """ID of the current table snapshot."""
+
+    snapshots: list
+    """A list of valid snapshots. Valid snapshots are snapshots for which
+    all data files exist in the file system. A data file must not be
+    deleted from the file system until the last snapshot in which it was
+    listed is garbage collected."""
+
+    snapshot_log: list = Field(alias="snapshot-log")
+    """A list (optional) of timestamp and snapshot ID pairs that encodes
+    changes to the current snapshot for the table. Each time the
+    current-snapshot-id is changed, a new entry should be added with the
+    last-updated-ms and the new current-snapshot-id. When snapshots are
+    expired from the list of valid snapshots, all entries before a snapshot
+    that has expired should be removed."""
+
+    metadata_log: list = Field(alias="metadata-log")
+    """A list (optional) of timestamp and metadata file location pairs that
+    encodes changes to the previous metadata files for the table. Each time
+    a new metadata file is created, a new entry of the previous metadata
+    file location should be added to the list. Tables can be configured to
+    remove oldest metadata log entries and keep a fixed-size log of the most
+    recent entries after a commit."""
+
+    sort_orders: list = Field(alias="sort-orders")
+    """A list of sort orders, stored as full sort order objects."""
+
+    default_sort_order_id: int = Field(alias="default-sort-order-id")
+    """Default sort order id of the table. Note that this could be used by
+    writers, but is not used when reading because reads use the specs stored
+     in manifest files."""
+
+
+class TableMetadataV1(TableMetadataCommonFields, IcebergBaseModel):
+    @root_validator(pre=True)
+    def populate_schemas(cls, data: Dict[str, Any]):
+        # When we read a V1 format-version, we'll make sure to populate the fields
+        # for V2 as well. This makes it easier downstream because we can just
+        # assume that everything is a TableMetadataV2.
+        # When writing, we should stick to the same version that it was,
+        # because bumping the version should be an explicit operation that is up
+        # to the owner of the table.
+        schema = data["schema"]
+        if "schemas" in data:
```

Comment:
I think you need the same handling for `partition-spec` and `partition-specs` / `default-spec-id` and some handling to initialize `sort-orders` / `default-sort-order-id`. You might want to check the spec and make sure this is correctly handling all of the optional v1 fields.

Reply:
Hey @rdblue I've added tests for handling the sort-order, based on this input:

- https://github.com/apache/iceberg/blob/master/core/src/test/resources/TableMetadataV1Valid.json
- https://github.com/apache/iceberg/blob/master/core/src/test/resources/TableMetadataV2Valid.json

This also shows that my previous test data was incorrect. For example, the partition spec doesn't have a spec id. I now set those to default for V1. I do believe that sort orders should work now, there are no fields to be migrated in V1/V2 if I understand correctly, it is just that it was optional before. We can also handle this as soon as we start adding the sort order classes. Let me know if I'm overlooking something here.